homeassistant/components/tasmota/device_trigger.py

```diff
@@ -0,0 +1,287 @@
+"""Provides device automations for MQTT."""
+import logging
+from typing import Callable, List, Optional
+
+import attr
+from hatasmota.trigger import TasmotaTrigger
+import voluptuous as vol
+
+from homeassistant.components.automation import AutomationActionType
+from homeassistant.components.device_automation import TRIGGER_BASE_SCHEMA
+from homeassistant.components.homeassistant.triggers import event as event_trigger
+from homeassistant.const import CONF_DEVICE_ID, CONF_DOMAIN, CONF_PLATFORM, CONF_TYPE
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.device_registry import CONNECTION_NETWORK_MAC
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from .const import DOMAIN, TASMOTA_EVENT
+from .discovery import TASMOTA_DISCOVERY_ENTITY_UPDATED, clear_discovery_hash
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_DISCOVERY_ID = "discovery_id"
+CONF_SUBTYPE = "subtype"
+DEVICE = "device"
+
+TRIGGER_SCHEMA = TRIGGER_BASE_SCHEMA.extend(
+    {
+        vol.Required(CONF_PLATFORM): DEVICE,
+        vol.Required(CONF_DOMAIN): DOMAIN,
+        vol.Required(CONF_DEVICE_ID): str,
+        vol.Required(CONF_DISCOVERY_ID): str,
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    }
+)
+
+DEVICE_TRIGGERS = "tasmota_device_triggers"
+
+
+@attr.s(slots=True)
+class TriggerInstance:
+    """Attached trigger settings."""
+
+    action: AutomationActionType = attr.ib()
+    automation_info: dict = attr.ib()
+    trigger: "Trigger" = attr.ib()
+    remove: Optional[CALLBACK_TYPE] = attr.ib(default=None)
+
+    async def async_attach_trigger(self):
+        """Attach event trigger."""
+        event_config = {
+            event_trigger.CONF_PLATFORM: "event",
+            event_trigger.CONF_EVENT_TYPE: TASMOTA_EVENT,
+            event_trigger.CONF_EVENT_DATA: {
+                "mac": self.trigger.tasmota_trigger.cfg.mac,
+                "source": self.trigger.tasmota_trigger.cfg.source,
+                "event": self.trigger.tasmota_trigger.cfg.event,
+            },
+        }
+
+        event_config = event_trigger.TRIGGER_SCHEMA(event_config)
+        if self.remove:
+            self.remove()
+        self.remove = await event_trigger.async_attach_trigger(
```

Comment:
Could this get issues with race condition?

Reply:
`event_trigger.async_attach_trigger` is actually a syncronous function, I added a comment instead of a lock.
It's a little bit weird that `event_trigger.async_attach_trigger` is declared async though if it's not needed? Maybe that should be fixed in another PR?